src/perf_counters.cc

```diff
@@ -128,5 +129,33 @@ PerfCounters PerfCounters::Create(
 
 PerfCounters::~PerfCounters() = default;
 #endif  // defined HAVE_LIBPFM
+
+int PerfCountersMeasurement::ref_count_ = 0;
+std::unique_ptr<PerfCounters> PerfCountersMeasurement::counters_;
+
+PerfCountersMeasurement::PerfCountersMeasurement(
+    const std::vector<std::string>& counter_names)
+    : start_values_(counter_names.size()),
+      end_values_(counter_names.size()) {
+  {
+    // TODO: Add MutexLock
+    //MutexLock l(mutex_);
+    if (!counters_) {
+      PerfCounters perf_counters = PerfCounters::Create(counter_names);
+      counters_ = std::unique_ptr<PerfCounters> (
+          new PerfCounters(std::move(perf_counters)));
```

Comment:
This is can be simplified if PerfCounters::Create returns a std::unique_ptr; OR come to think of it, I think we wanted PerfCounters to be allocatable inlined (i.e. on the stack, or as a field of a class/struct), so we'd avoid a dereference and added noise to measurements.

So let's have `counters_` be a PerfCounters object, invalid initially (add a default ctor for `PerfCounters` to do that) and a `reset()` function on it that does what the dtor does today (the dtor would call it, too, for completeness). The design is logically the same.

Reply:
When I tried to make `counters_` be a PerfCounters object, here is a problem.
The `counters_` is constructed firstly, and the `counter_names` is passed in later.
How about add move assignment to `PerfCounters`, so we could use `PerfCounters::Create()` and `PerfCounters::NoCounters()` without any modification.

Reply:
By all means - add what you need.